### PR TITLE
Added new article on writing good python code

### PR DESCRIPTION
**Pull request by ivanleomk**

### CODE REVIEW

File: app/blog/posts/good-python-code.mdx [hunk 1]
```diff
@@ -0,0 +1,333 @@
+---
+title: "Everything I've learnt about writing good Python code"
+publishedAt: "2024-04-17"
+description: Speedrun your way to becoming a good python developer and don't make the same mistakes I did
+---
+
+I've always used python for my interviews but it wasn't until about 6 months ago that I started to use python for all sorts of projects - from simple machine learning scripts to full blown applications. In the proccess, I've learnt a lot about what constitutes good python code and I hope that you'll find the tips in this article useful in your own projects.
```
Comment (jxnl) [inline app/blog/posts/good-python-code.mdx]: can definitely use a stronger hook. 
Today I'm going to share with you end tips on how to write better code that I learn from so on the thing.

File: app/blog/posts/good-python-code.mdx [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+
+I've been lucky enough to have Jason (@jxnlco on twitter) review a good chunk of my code and I've found that these few things have made a massive difference in my code quality.
+
+1. using the `@classmethod` decorator
```
Comment (jxnl) [inline app/blog/posts/good-python-code.mdx:9]:
```suggestion
1. then to use the `@classmethod` decorator
```

File: app/blog/posts/good-python-code.mdx [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+2. taking time to learn about common libraries
+3. writing simpler functions
```
Comment (jxnl) [inline app/blog/posts/good-python-code.mdx]:
```suggestion
3. write simpler functions
```

File: app/blog/posts/good-python-code.mdx [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+publishedAt: "2024-04-17"
+description: Speedrun your way to becoming a good python developer and don't make the same mistakes I did
+---
+
+I've always used python for my interviews but it wasn't until about 6 months ago that I started to use python for all sorts of projects - from simple machine learning scripts to full blown applications. In the proccess, I've learnt a lot about what constitutes good python code and I hope that you'll find the tips in this article useful in your own projects.
+
+I've been lucky enough to have Jason (@jxnlco on twitter) review a good chunk of my code and I've found that these few things have made a massive difference in my code quality.
+
+1. using the `@classmethod` decorator
+2. taking time to learn about common libraries
```
Comment (jxnl) [inline app/blog/posts/good-python-code.mdx]:
```suggestion
2. learn the stdlib
```

File: app/blog/posts/good-python-code.mdx [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+3. writing simpler functions
+4. being a bit lazier - earn the abstraction
+5. decouple your implementation
+
+## Use the classmethod decorator
```
Comment (jxnl) [inline app/blog/posts/good-python-code.mdx:15]: you cna call out the fact that instructor's new api looks like this 

from_opena
from_anthropic

File: app/blog/posts/good-python-code.mdx [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+````
+
+
+There are a few things which make the code above good
+
+1. **Easier To Read**: I can determine which provider I'm using when I instantiate the class - `EmbeddingModel.from_hf` makes it clear that it's the `SentenceTransformers` package that's being used
+
+2. **Lesser Redundancy**:  I only need to pass in the values that I need to for each specific model. This makes it easy to add in additional configuration parameters down the line and be confident that it won't mess up existing functionality
+
+## Learn Common Libraries
```
Comment (jxnl) [inline app/blog/posts/good-python-code.mdx:92]:
```suggestion
## Learn Standard Library. 
```

Comment (jxnl) [reply app/blog/posts/good-python-code.mdx:92]: Okay, I take that back. You do mean common libraries lol

let me also add stuff like 

`collections, functools, itertools`

File: app/blog/posts/good-python-code.mdx [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+        if s1 not in seen:
+            seen.add(s1)
+            yield s1
+
+        if s2 not in seen:
+            seen.add(s1)
+            yield s2
+
+
+def get_sentence_to_id_mapping(sentences: List[str]):
```
Comment (jxnl) [inline app/blog/posts/good-python-code.mdx:200]: looking again sn't this just 

`{s:i for i, s in enumerate(sentence)}`

File: app/blog/posts/good-python-code.mdx [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+    mapping = {}
+    for sentence in sentences:
+        mapping[sentence] = len(mapping)
+    return mapping
+
+
+def generate_sentence_batches(sentence_mapping: dict[str, int], batch_size=100):
```
Comment (jxnl) [inline app/blog/posts/good-python-code.mdx:204]: You should call out that an iter tools batch exists in Python 3.13

File: app/blog/posts/good-python-code.mdx [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+
+It helps to have one main function call a sequence of other functions and have those functions be as flat as possible. This means that ideally between each of these calls, we minimise mutation or usage of some shared variable and only do so when there's an expensive piece of computation involved.
+
+## Earn the Abstraction 
+
+I think it's easy to quickly complicate a codebase with premature abstractions without much effort over time - just look at Java. After all, it's a natural reflex as we work towards writing code that is DRY. But that often makes it difficult for you to adapt your code down the line. 
+
+For instance, an easy way to do this initially is to just return a simple dictionary if the returned value is only being used by a single function. 
+
+```
```
Comment (jxnl) [inline app/blog/posts/good-python-code.mdx]:
```suggestion
```python
```

File: app/blog/posts/good-python-code.mdx [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+    data = pd.read_csv("data.csv")
+    attributes = extract_attributes(data)
+    return attributes
+```
+
+## Decouple your implementation
+
+Another example I like a lot is scoring values. Say we want to calculate the recall for a list of predictions that we've made where we have a single known label as our ground truth and a list of other labels as our model's predictions. We might implement it down the line as 
+
+```
```
Comment (jxnl) [inline app/blog/posts/good-python-code.mdx]: fix all the python hints

File: app/blog/posts/good-python-code.mdx [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+
+In fact, we could even go one step further and just represent all of the metrics that we want as a single dictionary.
+
+```
+SIZES = [3, 10, 15, 25]
+metrics = {"recall": calculate_recall, "mrr": calculate_mrr}
+evals = {}
+
+
+def slice_predictions_at_k(k: int, score):
```
Comment (jxnl) [inline app/blog/posts/good-python-code.mdx]:
```suggestion
def predictions_at_k(score_fn, k:int):
```
the *fn makes it know its a function

k is a property of the s function k so it hould be second

k suggests slice

File: app/blog/posts/good-python-code.mdx [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+    def wrapper(chunk_id, predictions):
+        return score(chunk_id, predictions[:k])
+
+    return wrapper
+
+
+for metric, sz in itertools.product(metrics.keys(), SIZES):
```
Comment (jxnl) [inline app/blog/posts/good-python-code.mdx]: don't call it sz if we aggree its called k

k should be second arg

File: app/blog/posts/good-python-code.mdx [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+    evals[f"{metric}@{sz}"] = slice_predictions_at_k(sz, metrics[metric])
```
Comment (jxnl) [inline app/blog/posts/good-python-code.mdx]: avoid non keyword arguments

File: app/blog/posts/good-python-code.mdx (hunk 1 above)
Comment (ivanleomk) [reply app/blog/posts/good-python-code.mdx]: I changed it to 

> In the past 6 months, I've 10xed the amount of python code I've written. In this article, I'll show you a few easy actionable tips to write better and more maintanable code. I've been lucky enough to have Jason (@jxnlco on twitter) review a good chunk of my code and I've found that these few things have made a massive difference in my code quality.

File: app/blog/posts/good-python-code.mdx (hunk 7 above)
Comment (ivanleomk) [reply app/blog/posts/good-python-code.mdx:200]: You are indeed right haha

Comment (ivanleomk) [reply app/blog/posts/good-python-code.mdx:200]: Changed it to

```
def get_sentence_to_id_mapping(sentences: List[str]):
    return {s:i for i, s in enumerate(sentence)}
```

File: app/blog/posts/good-python-code.mdx (hunk 2 above)
Comment (ivanleomk) [reply app/blog/posts/good-python-code.mdx:9]: Hmm I think I prefer `using the class method decorator` in this case.

**Pull request merged**